File: coding-drills/tools/check_hardest.js
/* Regression test for the hardest-questions round (see tools/patch_hardest.py).
 *
 * Seeds a synthetic history, then asserts the behaviour that was asked for:
 *
 *   * the set is ranked by RAW wrong count and is the top 10% of everything
 *     attempted, with a floor so a round is never two questions;
 *   * a question never got wrong never appears, however often it was seen;
 *   * the round queues ONLY that set;
 *   * it IGNORES the topic and drill-mode pills -- the "global" choice;
 *   * the button is disabled, and says why, with no history;
 *   * an ordinary round afterwards still queues the full selection.
 *
 *     node tools/check_hardest.js <file>.html      # exits 1 on any failure
 */
const { JSDOM } = require('jsdom');
const fs = require('fs');

const file = process.argv[2];
const html = fs.readFileSync(file, 'utf-8');
const PROG_KEY = html.match(/PROG_KEY = "([^"]+)"/)[1];

function open(seed) {
  return new JSDOM(html, {
    runScripts: 'dangerously', pretendToBeVisual: true, url: 'https://localhost/',
    beforeParse(win) {
      if (seed) win.localStorage.setItem(PROG_KEY, JSON.stringify(seed));
    },
  });
}

const fails = [];
const check = (name, ok, detail) => {
  if (!ok) fails.push(name + (detail ? ' — ' + detail : ''));
};

/* ---- 1. no history: the button is there, disabled, and explains itself ---- */
const fresh = open(null);
const units = fresh.window.__pd.everyUnit();
{
  const btn = fresh.window.document.querySelector('#hardestBtn');
  check('button rendered on a fresh app', !!btn);
  if (btn) {
    check('button disabled with no history', btn.disabled);
    const note = btn.parentNode.querySelector('.small').textContent;
    check('fresh note names the threshold',
      /Unlocks once 8 different questions have been missed \(0 so far\)/.test(note), note);
  }
}
fresh.window.close();

/* ---- 2. seeded history: 120 attempted, 30 with mistakes (wrong 30..1), plus
        one question seen a great deal and never missed. ---- */
const seed = { __history: [] };
for (let i = 0; i < 120 && i < units.length; i++) {
  const wrong = i < 30 ? 30 - i : 0;
  seed[units[i].qid] = { seen: 5 + wrong, correct: 5, wrong: wrong, streak: 0 };
}
const masteredQid = units[119].qid;
seed[masteredQid] = { seen: 80, correct: 80, wrong: 0, streak: 9 };

const dom = open(seed);
const w = dom.window, d = w.document, pd = w.__pd;
check('seeded progress was loaded', Object.keys(pd.hardestUnits()).length > 0);

const hard = pd.hardestUnits();
const wrongs = hard.map(u => seed[u.qid].wrong);
check('pool is 10% of 120 attempted', hard.length === 12, 'got ' + hard.length);
check('ranked by raw wrong count',
  wrongs.every((v, i) => i === 0 || v <= wrongs[i - 1]), wrongs.join(','));
check('worst question is first', wrongs[0] === 30, String(wrongs[0]));
check('mastered question excluded', !hard.some(u => u.qid === masteredQid));
check('nothing unseen slipped in', hard.every(u => seed[u.qid] && seed[u.qid].seen));

/* ---- 3. the button, and a round that ignores the pills ---- */
const btn = d.querySelector('#hardestBtn');
check('button enabled once the pool is ready', btn && !btn.disabled);
check('button names the count', btn && /12/.test(btn.textContent), btn && btn.textContent);
check('note says it ignores the selections',
  btn && /ignores the topic and drill-mode selections/.test(
    btn.parentNode.querySelector('.small').textContent));

// narrow the selection hard: the round must ignore it
d.querySelector('[data-sel="topics-none"]').click();
[...d.querySelectorAll('[data-topic]')].find(p => !p.classList.contains('empty')).click();
const narrowed = pd.askableUnits().length;
check('selection really was narrowed', narrowed > 0 && narrowed < units.length,
  narrowed + ' of ' + units.length);

d.querySelector('#hardestBtn').click();
const eyebrow = d.querySelector('.eyebrow').textContent;
check('round labelled hardest', /hardest/.test(eyebrow), eyebrow);
check('round is 12 long, not the selection', /1\/12/.test(eyebrow), eyebrow);

const hardNames = new Set(hard.map(u => u.card.name));
const askedNames = new Set();
let guard = 500, asked = 0;
while (guard-- > 0) {
  if (d.querySelector('.card.summary')) break;
  const name = d.querySelector('.qname');
  if (name) { askedNames.add(name.textContent); asked++; }
  const next = [...d.querySelectorAll('button')].find(b => /^(next|finish)$/i.test(b.textContent.trim()));
  if (next) { next.click(); continue; }
  const choice = [...d.querySelectorAll('button.choice')].filter(b => !b.disabled)[0];
  if (choice) { choice.click(); continue; }
  const ans = d.querySelector('#ans'), submit = d.querySelector('#submit');
  if (ans && submit && !submit.disabled) { ans.value = 'zzz'; submit.click(); continue; }
  break;
}
const outside = [...askedNames].filter(n => !hardNames.has(n));
check('round terminated', !!d.querySelector('.card.summary'));
check('only hardest questions asked', outside.length === 0, outside.slice(0, 3).join(', '));
check('summary names the round', !!d.querySelector('.card.summary') &&
  /Hardest questions complete/.test(d.querySelector('.card.summary').textContent));

/* ---- 4. an ordinary round afterwards still queues the full selection ---- */
d.querySelector('#homeBtn').click();
const nSel = pd.askableUnits().length;
d.querySelector('#startBtn').click();
const eb2 = d.querySelector('.eyebrow').textContent;
check('ordinary round back to the selection',
  eb2.indexOf('/' + nSel) !== -1 && !/hardest/.test(eb2), eb2 + ' (selection ' + nSel + ')');

console.log(file.padEnd(20), 'pool=' + hard.length, 'asked=' + asked,
  'distinct=' + askedNames.size, 'outside=' + outside.length,
  fails.length ? 'FAIL' : 'ok');
if (fails.length) { fails.forEach(f => console.error('  - ' + f)); process.exit(1); }

File: language-trainers/scripts/check_hardest.py
#!/usr/bin/env python3
"""Regression test for the hardest-words round (see patch_hardest.py).

Seeds a synthetic history, then asserts the behaviour that was asked for:

  * the set is ranked by RAW wrong count and is the top 10% of everything
    attempted (with a floor so a round is never two questions);
  * a hardest round asks ONLY words from that set, over many draws;
  * it IGNORES the lesson and word-form filters — that is the "global" choice;
  * an ordinary round afterwards is unaffected;
  * words never got wrong never appear, however often they were seen.

    python3 scripts/check_hardest.py        # exits 1 on any failure
"""
import json, os, subprocess, sys, tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
LT = os.path.join(HERE, "..")
APPS = ["polish_trainer.html", "spanish_trainer.html", "italian_trainer.html",
        "french_trainer.html", "latin_trainer.html", "medical_trainer.html", "portuguese_trainer.html"]

PROBE = r"""
// Seed a history: 120 attempted words, 30 of them with mistakes, wrong counts
// spread so the ranking has something to sort. One word is seen very often but
// never missed -- it must never show up in a hardest round.
const ids = VOCAB_DATA.entries.map(function(e){ return e.id; });
progress.stats = {};
for (let i = 0; i < 120 && i < ids.length; i++) {
  const wrong = i < 30 ? (30 - i) : 0;          // 30..1 for the first thirty
  progress.stats[ids[i]] = { correct: 5, wrong: wrong, seen: 5 + wrong, streak: 0 };
}
progress.stats[ids[119]] = { correct: 80, wrong: 0, seen: 80, streak: 9 };  // mastered

const rows = hardestRows();
const set = {}; rows.forEach(function(r){ set[r.id] = true; });

// ranked by raw wrong count, descending
let ordered = true;
for (let i = 1; i < rows.length; i++) if (rows[i].wrong > rows[i-1].wrong) ordered = false;

// turn every drill type on so the round can always ask something
Object.keys(enabledModes).forEach(function(m){ enabledModes[m] = true; });

// Restrict the ordinary filters hard: a hardest round must ignore them.
const lessons = Object.keys(enabledLessons);
Object.keys(enabledLessons).forEach(function(l){ enabledLessons[l] = false; });
enabledLessons[lessons[0]] = true;

startHardestRound();
let asked = 0, outside = 0, distinct = {};
for (let i = 0; i < 600; i++) {
  const q = pickQuestion ? null : null;   // pickQuestion touches the DOM; use the generator
  const modes = Object.keys(enabledModes).filter(function(m){ return enabledModes[m]; });
  const mode = modes[i % modes.length];
  const c = generateQuestion(mode);
  if (!c) continue;
  asked++;
  distinct[c.entryId] = true;
  if (!set[c.entryId]) outside++;
}

// the header's selection line must describe the round, not the filters
const countEl = document.getElementById("selection-count");
renderSelectionCount();
const hardestCountLine = countEl.textContent;

// an ordinary round afterwards must go back to honouring the filters
startNewRound();
renderSelectionCount();
const normalCountLine = countEl.textContent;
let normalOutside = 0, normalAsked = 0;
for (let i = 0; i < 300; i++) {
  const modes = Object.keys(enabledModes).filter(function(m){ return enabledModes[m]; });
  const c = generateQuestion(modes[i % modes.length]);
  if (!c) continue;
  normalAsked++;
  const e = byId[c.entryId];
  if (e && !enabledLessons[e.lesson]) normalOutside++;
}

console.log(JSON.stringify({
  poolSize: rows.length,
  topWrong: rows.length ? rows[0].wrong : null,
  ordered: ordered,
  masteredIncluded: !!set[ids[119]],
  asked: asked,
  outsideSet: outside,
  distinctAsked: Object.keys(distinct).length,
  normalAsked: normalAsked,
  normalOutsideLesson: normalOutside,
  hardestCountLine: hardestCountLine,
  normalCountLine: normalCountLine,
  hardestFlagCleared: hardestMode === false
}));
"""


def run(app):
    src = open(os.path.join(LT, app), encoding="utf-8").read()
    js = src.split("<script>", 1)[1].rsplit("</script>", 1)[0]
    stub = open(os.path.join(HERE, "dom_stub.js"), encoding="utf-8").read()
    d = tempfile.mkdtemp()
    p = os.path.join(d, "p.js")
    open(p, "w", encoding="utf-8").write(stub + js + PROBE)
    r = subprocess.run(["node", p], capture_output=True, text=True)
    if r.returncode != 0:
        print(r.stderr[-1500:])
        return None
    return json.loads(r.stdout)


def main():
    fails = 0
    for app in APPS:
        res = run(app)
        if res is None:
            print(f"{app:24} FAILED to run")
            fails += 1
            continue
        checks = [
            ("pool is 10% of 120 attempted", res["poolSize"] == 12),
            ("ranked by wrong count", res["ordered"]),
            ("mastered word excluded", not res["masteredIncluded"]),
            ("questions produced", res["asked"] > 0),
            ("only hardest words asked", res["outsideSet"] == 0),
            ("ordinary round honours filters again", res["normalOutsideLesson"] == 0),
            ("flag cleared by new round", res["hardestFlagCleared"]),
            ("count line names the round",
             "hardest-words round" in res["hardestCountLine"]),
            ("count line back to selection",
             res["normalCountLine"].endswith("in selection")),
        ]
        bad = [n for n, ok in checks if not ok]
        status = "ok" if not bad else "FAIL: " + "; ".join(bad)
        print(f"{app:24} pool={res['poolSize']:3} asked={res['asked']:4} "
              f"distinct={res['distinctAsked']:3} outside={res['outsideSet']:3}  {status}")
        if bad:
            fails += 1
    print("\nall hardest-round checks pass" if not fails else f"\nFAILURES: {fails}")
    sys.exit(1 if fails else 0)


if __name__ == "__main__":
    main()
